userscript/src/components/edit-panel/lg-auto-turn-arrow/utils/set-control-turn-arrows.ts

```diff
@@ -0,0 +1,80 @@
+import { CountryDataModel } from '@/@waze/Waze/DataModels/CountryDataModel';
+import {
+  LaneGuidanceControlModel,
+  LaneGuidanceControlTurnModel,
+} from '@/components/edit-panel/lg-auto-turn-arrow/models';
+import { getWazeMapEditorWindow } from '@/utils/get-wme-window';
+import { getSegmentByVertex } from '@/utils/location';
+import { getSegmentHeadingByDirection } from '@/utils/wme-entities/segment';
+import { createVertexById } from '@/utils/wme-entities/segment-vertex';
+
+const arrowStyleStopPoints = [
+  { styleAngle: 135, stopPoint: 67.5 },
+  { styleAngle: 90, stopPoint: 112.5 },
+  { styleAngle: 45, stopPoint: 157.5 },
+  { styleAngle: 0, stopPoint: 202.5 },
+  { styleAngle: -45, stopPoint: 247.5 },
+  { styleAngle: -90, stopPoint: 292.5 },
+  { styleAngle: -135, stopPoint: 360 },
+];
+
+function isUTurn(turnModel: LaneGuidanceControlTurnModel): boolean {
+  const fromVertex = createVertexById(turnModel.get('fromVertexID'));
+  const toVertex = createVertexById(turnModel.get('toVertexID'));
+  if (fromVertex.getOpposite().equals(toVertex)) return true;
+
+  const dataModel = getWazeMapEditorWindow().W.model;
+  const fromSegment = getSegmentByVertex(dataModel, fromVertex);
+  const toSegment = getSegmentByVertex(dataModel, toVertex);
+  const fromSegmentAddress = fromSegment.getAddress().format();
+  const toSegmentAddress = toSegment.getAddress().format();
+  const isAddressEquals = fromSegmentAddress === toSegmentAddress;
+  const fromSegmentRoadType = fromSegment.getAttribute('roadType');
+  const toSegmentRoadType = toSegment.getAttribute('roadType');
+
+  if (fromSegmentRoadType === toSegmentRoadType) {
+    if (
+      (isAddressEquals || fromSegmentRoadType === toSegmentRoadType) &&
+      getSegmentHeadingByDirection(
+        fromSegment,
+        fromVertex.getOpposite().direction,
+      ) === getSegmentHeadingByDirection(toSegment, toVertex.direction)
+    ) {
+      return true;
+    }
+  }
+
+  return false;
```

Comment:
Note you've changed the logic, as the address might differ, but as long as the road type remains the same - we can consider the turn as a u-turn. I am fine with the existing implementation, although I agree it can be rewritten at some point later along with other similar functions.